lib/ruby_lsp/base_server.rb

```diff
@@ -31,6 +31,7 @@ def initialize(test_mode: false)
         Thread,
       )
 
+      @global_state = T.let(GlobalState.new, GlobalState)
```

Comment:
I had to move global state to the base server, so that we can notify the client about the virtual source content.